### PR TITLE
mvynokur: added Google test to verify search

### PR DESCRIPTION
**Pull request by MyroslavaVynokur**

### CODE REVIEW

File: src/test/java/mvynokur/GoogleTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static com.codeborne.selenide.Condition.*;
+import static com.codeborne.selenide.Selenide.*;
+
+public class GoogleTest {
+
+    @BeforeMethod
+    public void setUp() {
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:15]: rename

File: src/test/java/mvynokur/GoogleTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        open("https://www.google.com");
+        $(By.name("q")).val("funny dogs").pressEnter();
+    }
+
+    @Test
+    @Description("Task 1")
+    public void verify_firstLink_contains_dogs() {
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:22]: rename

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/mvynokur/GooglePage.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import static com.codeborne.selenide.Selenide.$;
+import static com.codeborne.selenide.Selenide.$x;
+
+public class GooglePage {
+
+    private final String LogoIcon = "//div[@class = 'logo']";
+    private final String PaginationBlock = "//tr[@jsname = 'TeSSVd']";
+    private final String PreviousPage = PaginationBlock + "//a/span[contains(text(), 'Previous')]";
+    private final String NextPage = PaginationBlock + "//a/span[contains(text(), 'Next')]";
+    private final String UkrainianLanguageLink = "//div[@id = 'SIvCob']/a[contains(text(), 'українська')]";
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:15]: no cyrillic in locators

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import com.codeborne.selenide.ElementsCollection;
+import com.codeborne.selenide.SelenideElement;
+
+import static com.codeborne.selenide.Selenide.*;
+
+public class SearchResultsPage {
+
+    private final String LinkName = "//h3[@class = 'LC20lb MBeuO DKV0Md']";
+
+    public ElementsCollection searchResultLinksNames() {
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:12]: rename

File: src/test/java/mvynokur/GooglePageBaseMethods.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package mvynokur;
+
+import org.openqa.selenium.By;
+import org.testng.annotations.BeforeMethod;
+import static com.codeborne.selenide.Selenide.*;
+
+public class GooglePageBaseMethods {
```
Comment (toleksyn) [inline src/test/java/mvynokur/GooglePageBaseMethods.java:7]: ???

File: src/test/java/mvynokur/GoogleTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        GooglePage googlePage = new GooglePage();
+        googlePage.navigateToPage(5);
+        SearchResultsPage results = new SearchResultsPage();
+        results.searchResultLinksNames().get(0).shouldHave(text("dogs"));
+    }
+
+    @Test
+    @Description("Task 5")
+    public void verifySearchResultsLinksAmount() {
+        SearchResultsPage results = new SearchResultsPage();
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:50]: no raw page objects in tests

File: src/main/java/mvynokur/GooglePage.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import static com.codeborne.selenide.Selenide.$x;
+
+public class GooglePage {
+
+    private final String logoIcon = "//div[@class = 'logo']";
+    private final String paginationBlock = "//tr[@jsname = 'TeSSVd']";
+    private final String previousPage = paginationBlock + "//a/span[contains(text(), 'Previous')]";
+    private final String nextPage = paginationBlock + "//a/span[contains(text(), 'Next')]";
+    private final String feelingLuckyButton = "//div[@class='FPdoLc lJ9FBc']//input[@name='btnI']";
+    private final String languageBlock = "//div[@id='SIvCob']";
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:16]: why are these fields if they are being used once?

File: src/main/java/mvynokur/GooglePage.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:17]: rename

File: src/main/java/mvynokur/GooglePage.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    public GooglePage searchFor(String text) {
+        $(By.name("q")).val(text).pressEnter();
+        return new GooglePage();
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:20]: no need to create new object

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package mvynokur;
+
+import com.codeborne.selenide.ElementsCollection;
+import com.codeborne.selenide.SelenideElement;
+
+import static com.codeborne.selenide.Selenide.*;
+
+public class SearchResultsPage {
+
+    private final String linkName = "//h3[@class = 'LC20lb MBeuO DKV0Md']";
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:10]: rename

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        return $$x(linkName);
+    }
+
+    public SelenideElement getLinkNameByNumber(int index) {
+        return $x("//div[" + index + "]" + linkName + "/parent::a");
+    }
+
+    public ElementsCollection getDescriptions() {
+        return $$x("//div[@class = 'Uroaid']");
+    }
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:22]: PO should only have action methods and methods, that return data

File: src/test/java/mvynokur/BaseTestRunner.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.testng.annotations.BeforeMethod;
+
+import static com.codeborne.selenide.Selenide.open;
+
+public class BaseTestRunner {
+
+    @BeforeMethod
+    public void openHomePage() {
+        open("https://www.google.com");
```
Comment (toleksyn) [inline src/test/java/mvynokur/BaseTestRunner.java:11]: breaks PO model

File: src/test/java/mvynokur/GoogleTest.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        new SearchResultsPage().getLinkNameByNumber(9).
+                shouldHave(attributeMatching("href", "((http|https)://)(www.)?"
+                        + "[a-zA-Z0-9@:%._\\+~#?&//=]{2,256}\\.[a-z]"
+                        + "{2,6}\\b([-a-zA-Z0-9@:%._\\+~#?&//=]*)"));
+
+    }
+
+    @Test
+    @Description("Task 3")
+    public void verifyNavigationToHomePage() throws InterruptedException {
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:33]: why is this throws needed?

Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:33]: why is this throws needed?

File: src/main/java/mvynokur/GooglePage.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public GooglePage logoIconIsVisible() {
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:23]: Java code convention: isLogoVisible() and should return a boolean

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/test/java/mvynokur/BaseTestRunner.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+public class BaseTestRunner {
+
+    protected GooglePage homePage;
+    protected SearchResultsPage searchResults;
+    private final String baseSearch = "funny dogs";
+
+    @BeforeMethod
+    public void openHomePageAndSearch() {
+        homePage = new GooglePage().openHomePage().searchFor(baseSearch);
+        searchResults = new SearchResultsPage();
```
Comment (dankomax) [inline src/test/java/mvynokur/BaseTestRunner.java:14]: Should be searchResultsPage = new SearchResultsPage();

File: src/main/java/mvynokur/GooglePage.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+    public GooglePage searchFor(String text) {
+        $(By.name("q")).val(text).pressEnter();
+        return this;
+    }
+
+    public boolean isLogoVisible() {
+        try {
+            return $x(logoIconXpath).isDisplayed();
+        } catch (ElementNotFound e) {
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:28]: does this really work?

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
+package mvynokur;
+
+import static com.codeborne.selenide.Selenide.*;
+
+public class SearchResultsPage {
+
+    private final String headerXpath = "//h3[@class = 'LC20lb MBeuO DKV0Md']";
+
+    public int amountOfLinks() {
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:9]: rename, get*()

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+        return $$x(headerXpath).size();
+    }
+
+    public String getLinkNameNumber(int number) {
+        return $x("//div[" + number + "]" + headerXpath + "/parent::a").text().toLowerCase();
+    }
+
+    public String getDescriptionTextByLinkNumber(int number) {
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:17]: rename, getDescriptionText(int linkNumber)

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/mvynokur/GooglePage.java [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.codeborne.selenide.Selenide;
+import mvynokur.utils.WebElementUtils;
+import org.openqa.selenium.By;
+
+import static com.codeborne.selenide.Selenide.$;
+
+public class GooglePage {
+
+    public GooglePage openHomePage() {
```
Comment (toleksyn) [inline src/main/java/mvynokur/GooglePage.java:11]: rename

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/test/java/mvynokur/GoogleTest.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Test
+    @Description("Task 1")
+    public void verifySearchFunctionalityWorks() {
+        Assert.assertTrue(searchResultsPage.getName(1).contains("dogs"));
+    }
+
+    @Test
+    @Description("Task 2")
+    public void verifyNineLinkURLCorrectness() {
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:18]: camel case

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/test/java/mvynokur/BaseTestRunner.java [hunk 21]
```diff
@@ -0,0 +1,14 @@
+package mvynokur;
+
+import org.testng.annotations.BeforeMethod;
+
+public class BaseTestRunner {
+
+    protected GooglePage homePage = new GooglePage();
+    protected SearchResultsPage searchResultsPage = new SearchResultsPage();
```
Comment (toleksyn) [inline src/test/java/mvynokur/BaseTestRunner.java:8]: you should get this from a PO method call

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/mvynokur/SearchResultsPage.java [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    public String getName(int linkNumber) {
+        return $x(getLink(linkNumber)).text().toLowerCase();
+    }
+
+    public String getUrl(int linkNumber) {
+        return $x(getLink(linkNumber)).getAttribute("href");
+    }
+
+    public String getDescriptionText(int linkNumber) {
+        return $$x("//div[@class = 'Uroaid']").get(linkNumber - 1).text().toLowerCase();
```
Comment (toleksyn) [inline src/main/java/mvynokur/SearchResultsPage.java:50]: use direct locator

File: src/test/java/mvynokur/GoogleTest.java [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    @Test
+    @Description("Task 5")
+    public void verifySearchResultsLinksAmount() {
+        Assert.assertTrue(searchResultsPage.getLinksAmount() >= 9);
+    }
+
+    @Test
+    @Description("Task 6/7")
+    public void verifySearchFunctionalityWithDifferentInputs() {
+        searchResultsPage.clearSearchField().searchFor("funny kitten");
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:52]: format properly

File: src/test/java/mvynokur/GoogleTest.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        SoftAssert softAssert = new SoftAssert();
+        softAssert.assertTrue(searchResultsPage.getDescriptionText(1).contains("kitten"));
```
Comment (toleksyn) [inline src/test/java/mvynokur/GoogleTest.java:54]: searchResultsPage.getDescriptionText(1) - extract to var, re-use

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Pull request merged**